### PR TITLE
Pasting near a table grows the table to the left and down

### PR DESCRIPTION
**Pull request by ddimaria**
## Relevant issue(s)
Fixes https://github.com/quadratichq/quadratic/issues/2624

## Tasks
- [x] Extract out growing logic from set_cells in order to be reused
- [x] Apply table growing to pasting
- [x] Write tests 

## Description
Pasting to end of table (bottom or right) doesn't expand table.  This PR expands to the bottom or to the right when pasted directly to the bottom or right.  Apply the operations in 2 separate transactions so that the user can undo the table growing without undoing the paste.

Some exceptions:
* No expansion if the paste is touching table heading (would cause a spill otherwise)
* Expansions don't happen when within a data table
* Expansions don't happen grid data would be overwritten by the paste (for the overwritten portion only)
* Tables don't expand over other tables (they stop at the edge)

## Testing considerations
There were a lot of edge cases.  I tried to capture them all but please test all-the-things.

## QA Wolf tests 
QA Wolf should add tests for this new functionality.

## Demo
https://github.com/user-attachments/assets/d3800dc6-e5a0-497c-8f68-78108c9e9b95

### CODE REVIEW

**Comment by luke-quadratic** [comment]
Redeploy

File: quadratic-core/src/grid/sheet/data_table.rs [hunk 1]
```diff
@@ -331,18 +331,33 @@ impl Sheet {
...
+
+        let data_table_immediate_left = data_tables
+            .iter()
+            .find(|rect| rect.intersects(rect_moved_left));
+
+        let should_not_expand = data_table_immediate_left.map(|rect| rect_moved_left.min.y <= rect.min.y)
+            .unwrap_or(false);
+
+        !should_not_expand
+    }
```
Comment (AyushAgrawal-A2) [inline quadratic-core/src/grid/sheet/data_table.rs:347]: one edge case because of this logic, not sure of if important or we want the remaining table to grow

File: quadratic-core/src/controller/operations/data_table.rs [hunk 2]
```diff
@@ -292,6 +294,107 @@ impl GridController {
...
+                        rect.max.y += 1;
+                    });
+            }
+        }
+    }
+
+    /// Returns operations to grow a data table to the right or bottom if the
+    /// cell value is touching the right or bottom edge.
+    pub fn grow_data_table_operations(
+        &self,
```
Comment (AyushAgrawal-A2) [inline quadratic-core/src/controller/operations/data_table.rs:364]: `&self` can be removed

File: quadratic-core/src/controller/operations/clipboard.rs [hunk 3]
```diff
@@ -649,6 +661,14 @@ impl GridController {
...
+        // collect information for growing data tables
+        let mut data_table_columns: HashMap<SheetPos, Vec<u32>> = HashMap::new();
+        let mut data_table_rows: HashMap<SheetPos, Vec<u32>> = HashMap::new();
+        let existing_data_tables = self.a1_context_sheet_table_bounds(sheet_id);
+        let mut growing_data_tables = existing_data_tables.clone();
+        let bypass_growing =
+            !Sheet::should_expand_data_table(&existing_data_tables, clipboard_rect);
+
         // loop through the clipboard and replace cell references in formulas and other languages
         for (start_x, x) in (insert_at.x..=max_x)
```
Comment (AyushAgrawal-A2) [inline quadratic-core/src/controller/operations/clipboard.rs:673]: update comment pls, we are checking for grow table also now

File: quadratic-core/src/controller/operations/clipboard.rs [hunk 4]
```diff
@@ -692,11 +707,34 @@ impl GridController {
...
                                     adjust,
                                 );
+                            } else {
+                                let new_x = x + cols_x as i64;
+                                let new_y = y + *cols_y as i64;
+                                let current_pos = Pos::new(new_x, new_y);
+                                let within_data_table = existing_data_tables
+                                    .iter()
+                                    .any(|rect| rect.contains(current_pos));
+
```
Comment (AyushAgrawal-A2) [inline quadratic-core/src/controller/operations/clipboard.rs:722]: this part seems expensive, we are 3 loops deep here and then iterate over all data tables (Imports in this sheet), we can filter `existing_data_tables` to be just the ones intersecting with clipboard rect to reduce this loop size to only our region of interest

File: quadratic-core/src/controller/operations/cell_value.rs [hunk 5]
```diff
@@ -30,14 +30,7 @@ impl GridController {
...
         let mut data_table_ops = vec![];
-        let existing_data_tables = self
-            .a1_context()
-            .tables()
-            .filter(|table| {
-                table.sheet_id == sheet_pos.sheet_id && table.language == CodeCellLanguage::Import
-            })
-            .map(|table| table.bounds)
-            .collect::<Vec<_>>();
+        let existing_data_tables = self.a1_context_sheet_table_bounds(sheet_pos.sheet_id);
```
Comment (AyushAgrawal-A2) [inline quadratic-core/src/controller/operations/cell_value.rs:33]: same as prev, comment, we can filter `existing_data_tables` to be just the ones intersecting with `values` rect to reduce the loop size later, which is 2 loops deep

**Review by AyushAgrawal-A2 — CHANGES_REQUESTED** [review]
few comments and 1 test failing

test failing as now table is growing on paste, the test assertion needs to be updated

**Review by luke-quadratic — APPROVED** [review]
approved on functionality - my discussion item around pasting empty/formatted was addressed in standup 👍

File: quadratic-core/src/controller/operations/clipboard.rs (hunk 3 above)
Comment (ddimaria) [reply quadratic-core/src/controller/operations/clipboard.rs:673]: Done

File: quadratic-core/src/controller/operations/cell_value.rs (hunk 5 above)
Comment (ddimaria) [reply quadratic-core/src/controller/operations/cell_value.rs]: Done

**Review by AyushAgrawal-A2 — APPROVED** [review]
